### PR TITLE
LibraryPanels: Load library panels in the frontend rather than the backend

### PR DESCRIPTION
**Pull request by ryantxu**
This PR updates the library panel loading logic so that we combine panels into the dashboard in frontend code rather than backend code.  This change will:
1. greatly simplify the backend loading logic
2. allows library panels to use the same schema upgrade logic that is used for dashboards
3. simplifies the security constraints for dashboards that *may* load a library panel

### CODE REVIEW

File: public/app/features/library-panels/state/api.ts [hunk 1]
```diff
@@ -76,9 +89,9 @@ export async function addLibraryPanel(
 }
 
 export async function updateLibraryPanel(panelSaveModel: PanelModelWithLibraryPanel): Promise<LibraryElementDTO> {
-  const { uid, name, version } = panelSaveModel.libraryPanel;
+  const { libraryPanel, ...model } = panelSaveModel;
```
Comment (ryantxu) [inline public/app/features/library-panels/state/api.ts:92]: currently library panels are getting saved with the `libraryPanel` element embedded in them.  This removes it before saving

**Review by torkelo — COMMENTED** [review]
Don't know how you can start so many things like this at the same time! :heart:  

I have been thinking of taking a stab at this myself.  Because I really regret the decision me and Hugo came to (I am to blame mostly) to load library panels in the backend, and clean in backend. Just plain stupid and more complex solution. 

I tested this and it's working really well. I was worried at first that doing async after dashboard has loaded and panels start rendering was too late. And the way you first load fake panel plugin during library panel load would case issues with double calls to panel.pluginLoaded and the mutable state of PanelModel would be in a strange place. But see that this actually does not happen, and that initPanelState is not fully run during loading. 

So think this will work really well. And even better with the new Scene runtime model where the messy mutable update after library panel loads will be a plain panel.setState call (hopefully) 

This should be able to get rid of a lof of backend code :) 

pushed a small update to fix an issue with the Dashboard exporter (still one test failing, for library panels inside collapsed rows)

**Comment by ryantxu** [comment]
> Don't know how you can start so many things like this at the same time!

🤣 believe it or not... related... the current library panel approach is a big complication for a simple entity store.  I needed to prove to myself that we could do it in the frontend so the backend can be much easier/dumb.

**Comment by torkelo** [comment]
Started on a bigger refactoring of making panel.libraryPanel always just be a ref object (with name and uid) instead of containing the loaded library element DTO as well. But that got a bit to big so saved that in a branch and started again, but did a small type cleaning up panel.libraryPanel. 

Anyway we have two remaining issues
* Does not work on initial load of dashboard. For me all the library panels contained "testdata" default query after initital dashboard load. Was only after refresh the right queries where shown (I tried without any of my changes so it's not something I introduced). The problem is that the LoaderView plugin did not have skipDataQuery and my default data source is testdata and it executed a default query (panels have a default empty query as the default state). So after the library panel is loaded it just renders with that old query data. I added skipDataQuery to the LoaderView plugin, this fixes the problem but not all panels say "No data" after initial dashboard load. 
* The dashboard exporter needs some bigger changes, because of the lazy loading of library panels all models will not be loaded. So we need to add a step to the exported that loads all library elements. 

Will try to continue work in a day or two but need to continue on plugin page & topnav a bit

File: pkg/api/dashboard.go [hunk 2]
```diff
@@ -214,11 +214,11 @@ func (hs *HTTPServer) GetDashboard(c *models.ReqContext) response.Response {
 	// make sure db version is in sync with json model version
 	dash.Data.Set("version", dash.Version)
 
-	// load library panels JSON for this dashboard
-	err = hs.LibraryPanelService.LoadLibraryPanelsForDashboard(c.Req.Context(), dash)
-	if err != nil {
-		return response.Error(500, "Error while loading library panels", err)
-	}
+	// // load library panels JSON for this dashboard
+	// err = hs.LibraryPanelService.LoadLibraryPanelsForDashboard(c.Req.Context(), dash)
```
Comment (ryantxu) [inline pkg/api/dashboard.go:218]: @torkelo -- do you think we should remove the backend code as part of this PR?

**Review by torkelo — APPROVED** [review]
Took this for another spin, could not find any issues :)

**Comment by spinillos** [comment]
@ryantxu @torkelo We were using [LoadLibraryPanelsForDashboard](https://github.com/grafana/grafana/pull/50560/files#diff-2800bdf355459d293ccc0508973203bb42adefff53bbbca8d87190fc4807a9e9L54) function to create the CSV in reports [here](https://github.com/grafana/grafana-enterprise/blob/main/src/pkg/extensions/report/render/render.go#L281). 

What is the alternative of this?

**Comment by torkelo** [comment]
Regarding the csv report usage it was fixed in https://github.com/grafana/grafana-enterprise/pull/4070

**Pull request merged**